velox/functions/prestosql/types/IPAddressType.cpp

```diff
@@ -0,0 +1,216 @@
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  bool isSupportedToType(const TypePtr& other) const override {
+    switch (other->kind()) {
+      case TypeKind::VARBINARY:
+        return true;
+      case TypeKind::VARCHAR:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  void castTo(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (input.typeKind() == TypeKind::VARCHAR) {
+      castFromString(input, context, rows, *result);
+    } else if (input.typeKind() == TypeKind::VARBINARY) {
+      castFromVarbinary(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from {} to IPAddress not supported", resultType->toString());
+    }
+  }
+
+  void castFrom(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (resultType->kind() == TypeKind::VARCHAR) {
+      castToString(input, context, rows, *result);
+    } else if (resultType->kind() == TypeKind::VARBINARY) {
+      castToVarbinary(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from IPAddress to {} not supported", resultType->toString());
+    }
+  }
+
+ private:
+  static void castToString(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipaddresses = input.as<SimpleVector<int128_t>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      const auto intAddr = ipaddresses->valueAt(row);
+      folly::ByteArray16 addrBytes;
+      std::string s;
+      memcpy(&addrBytes, &intAddr, kIPAddressBytes);
+
+      bigEndianByteArray(addrBytes);
+      folly::IPAddressV6 v6Addr(addrBytes);
+
+      if (v6Addr.isIPv4Mapped()) {
+        s = v6Addr.createIPv4().str();
+      } else {
+        s = v6Addr.str();
+      }
+
+      exec::StringWriter<false> result(flatResult, row);
+      result.append(s);
+      result.finalize();
+    });
+  }
+
+  static void castFromString(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<int128_t>>();
+    const auto* ipAddressStrings = input.as<SimpleVector<StringView>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      const auto ipAddressString = ipAddressStrings->valueAt(row);
+      int128_t intAddr;
+      folly::IPAddress addr(ipAddressString);
+
+      auto addrBytes = folly::IPAddress::createIPv6(addr).toByteArray();
```

Comment:
Does this throw? Can we use a non-throwing API?

Reply:
It does throw but I don't think it could ever throw in our case. Here is the code excerpt from folly for the lower call of createIPv6. 

https://github.com/facebook/folly/blob/dadf30d52e249512e31de89aa30cced7e9638865/folly/IPAddress.cpp#L65
```
IPAddressV6 IPAddress::createIPv6(const IPAddress& addr) {
  if (addr.isV6()) {
    return addr.asV6();
  } else {
    return addr.asV4().createIPv6();
  }
}
```

https://github.com/facebook/folly/blob/dadf30d52e249512e31de89aa30cced7e9638865/folly/IPAddress.h#L289C1-L300C4
```
  /**
   * Converts an IPAddress to an IPAddressV6 instance
   *
   * @throws InvalidAddressFamilyException if the IP Address is not currently
   * a valid V6 instance.
   */
  const IPAddressV6& asV6() const {
    if (FOLLY_UNLIKELY(!isV6())) {
      asV6Throw();
    }
    return addr_.ipV6Addr;
  }
```

It will throw if not a valid v6 instance. However if the string is valid IP it would fall under IPV4 or IPV6. If it was not a valid IP we would have thrown an error during the parsing of the constructor for `folly::IPAddress addr(ipAddressString);`

Reply:
> If it was not a valid IP we would have thrown an error during the parsing of the constructor for folly::IPAddress addr(ipAddressString);

We need to find a way to avoid throwing. Perhaps, we could use folly::IPAddress::tryFromString.

Reply:
Copied from here.
https://github.com/facebook/folly/blob/2f7582dc07152c36ebd3959e9fd424ec2fb7667c/folly/IPAddress.cpp#L221
Removed the throw and replaced with setStatus.